rest-service/manager_rest/rest/resources_v3_1/tokens.py

```diff
@@ -1,35 +1,110 @@
+    def post(self):
+        """Create a new token."""
+        request_dict = get_json_and_verify_params({
+            'description': {'type': str},
+            'expiration_date': {'optional': True},
+        })
+
         sm = get_storage_manager()
-        user = sm.get(models.User, user_id)
-        token = user.get_auth_token()
-        return dict(username=user.username,
-                    value=token, role=user.role)
+
+        secret = _random_string(40)
+
+        expiration_date = request_dict.get('expiration_date')
+        if expiration_date:
+            expiration_date = parse_utc_datetime(expiration_date)
+            if is_expired(expiration_date):
+                raise BadParametersError("Expiration date was in the past.")
+
+        token = models.Token(
+            id=_random_string(),
+            description=request_dict.get('description', ''),
+            secret_hash=hash_password(secret),
+            expiration_date=expiration_date,
+            _user_fk=current_user.id,
+        )
+        sm.put(token)
+
+        # Keeping the same structure as old tokens to make migration easier,
+        # but adding relevant extra details.
+        return dict(username=current_user.username,
+                    value=f'ctok-{token.id}-{secret}',
+                    role=current_user.role,
+                    expiration_date=token.expiration_date,
+                    last_used=token.last_used,
+                    token_id=token.id)
+
+
+class TokensId(SecuredResource):
+    @authorize('delete_token')
+    def delete(self, token_id):
+        """Delete an existing token."""
+        sm = get_storage_manager()
+        token = sm.get(models.Token, token_id, fail_silently=True)
+        if token and _can_manage_token(token):
+            sm.delete(token)
+            return None, 204
+        else:
+            raise NotFoundError()
+
+    @marshal_with(responses.Tokens)
+    @authorize('list_token')
+    def get(self, token_id):
+        sm = get_storage_manager()
+        token = sm.get(models.Token, token_id, fail_silently=True)
+
+        if token and _can_manage_token(token):
+            token_user = sm.get(models.User, token._user_fk)
+            return dict(username=token_user.username,
+                        value=f'ctok-{token.id}-********',
+                        role=token_user.role,
+                        expiration_date=token.expiration_date,
+                        last_used=token.last_used,
+                        token_id=token.id)
+        else:
+            raise NotFoundError()
+
+
+def _can_manage_token(token):
+    if token._user_fk == current_user.id:
+        return True
+    elif is_user_action_allowed('manage_others_tokens'):
+        return True
+    return False
+
+
+# A bit more user friendly than uuids
+def _random_string(length=10):
+    charset = string.ascii_uppercase + string.ascii_lowercase + string.digits
+    return ''.join(random.choice(charset) for i in range(length))
```

Comment:
this is also used for the token secrets, so I think just out of principle, I'd use SystemRandom here.

And actually, we're on 3.6, so:
```suggestion
    return ''.join(secrets.choice(charset) for i in range(length))
```
(and import secrets)

Reply:
Updating.